### PR TITLE
ELY-2308: Fix digest authentication for requests with encoded query parameters

### PR DESCRIPTION
**Pull request by hpehl**
https://issues.redhat.com/browse/ELY-2308

### CODE REVIEW

**Comment by hpehl** [comment]
I was not sure how to calculate the nonces in `org.wildfly.security.http.digest.DigestAuthenticationMechanismTest`, so I didn't provide an additional test case for the fix.

**Comment by fjuma** [comment]
Thanks @hpehl! The fix looks good but we should try to add a test if possible. In the current test class, the nonces were taken from examples in the spec (https://datatracker.ietf.org/doc/html/rfc7616) but there isn't an example with an encoded query param in there.

I think you should still be to create a new test based off one of the existing ones and tweak some of the content.

You'll also need to set `org.wildfly.security.http.validate-digest-uri` to `true` in the test in order to make sure `DigestAuthenticationMechanism#digestUriMatchesRequestUri` is called.

**Comment by hpehl** [comment]
> I think you should still be to create a new test based off one of the existing ones and tweak some of the content.

Sorry @fjuma, but I have no idea how to calculate the nounces. Sure, I can create a new test based on one of the existing tests like `testRfc2617()`, but as soon as I change the request I need to provide a matching response. 

Any suggestions, help, links, or guides on how to calculate the nounces and any other necessary data is highly appreciated.

**Comment by skyllarr** [comment]
@hpehl Yes it is a little complicated.  Nonce does not have to be changed, it is preconfigured for the tests. Previously we used values from RFC documentation, so there is no method that computes the digest in the tests. I think for your test case, you only need to change the URI to have encoded query parameter.

[Here](https://github.com/Skyllarr/wildfly-elytron/blob/ELY-2308/tests%2Fbase%2Fsrc%2Ftest%2Fjava%2Forg%2Fwildfly%2Fsecurity%2Fhttp%2Fdigest%2FDigestAuthenticationMechanismTest.java#L96) I uploaded the changes that show how the response field is calculated for `testRfc2617` method. You can use that approach to compute your [response](https://github.com/Skyllarr/wildfly-elytron/blob/ELY-2308/tests%2Fbase%2Fsrc%2Ftest%2Fjava%2Forg%2Fwildfly%2Fsecurity%2Fhttp%2Fdigest%2FDigestAuthenticationMechanismTest.java#L88) field and hard code it to your new test. Let me know if this helped.

Since digest computation can have different arguments/algorithms to consider, it would be a separate issue to refactor the tests to not use hard coded values.

**Comment by hpehl** [comment]
@Skyllarr thanks for your code. I've used it to add a test for a URL with an encoded query parameter. Please have a look and let me know what you think.

**Comment by skyllarr** [comment]
@hpehl Perfect! Test looks good, only do you mind removing the helper methods `encode` and `computeDigest` and just use the resulting value `9f4c595c275c13d41d65043d3a509abe` ? So the response parameter in your test would be: `response=\"9f4c595c275c13d41d65043d3a509abe\",\n" +`.

Because those helper methods would be a separate issue and would need to be double checked an refactored. I will create new jira issue for that. Thank you!

**Comment by hpehl** [comment]
Ok, done!

**Review by skyllarr — APPROVED** [review]

**Comment by skyllarr** [comment]
@hpehl Thank you! I created the following issue https://issues.redhat.com/browse/ELY-2310 to track the enhancement of this test class

**Comment by fjuma** [comment]
Thanks @Skyllarr and @hpehl!

@hpehl If you could squash the last two commits, that would be great.

**Comment by hpehl** [comment]
@fjuma Squashed my last two commits.

**Review by fjuma — APPROVED** [review]
Perfect, thanks @hpehl!

File: http/digest/src/main/java/org/wildfly/security/http/digest/DigestAuthenticationMechanism.java [hunk 1]
```diff
@@ -282,7 +282,7 @@ private boolean digestUriMatchesRequestUri(HttpServerRequest request, byte[] dig
             if (query == null || query.isEmpty()) {
                 relativeRequestUri = requestURI.getPath();
             } else {
-                relativeRequestUri = requestURI.getPath() + "?" + requestURI.getQuery();
+                relativeRequestUri = requestURI.getPath() + "?" + requestURI.getRawQuery();
```
Comment (gjaekel) [inline http/digest/src/main/java/org/wildfly/security/http/digest/DigestAuthenticationMechanism.java:285]: One must also use `getRawPath()` here to solve the Issue for the path of the URL, also.

**Pull request merged**